allow TestMo resources

derekk-nm opened the pull request:
# SUMMARY:
add support for creating TestMo "resource" files and allow passing one to the testmo-run-submit-thread action.
This is to enable reporting resources when submitting LM-Eval run results to TestMo.  Explicit testing of that will verify that we get the expected behavior when resource file is included.

# TEST PLAN:
Watch that the remote push for this branch succeeds, to verify the current case where no resources are being included in a run's thread.
I also verified that the essence of the `bash` worked from a terminal:

## looping through an associative array in the action that creates the resource file
```
root:~# declare -A fields=(["python"]="inputs.python_version" ["runner"]="inputs.runner" ["lm_eval_configuration"]="")
root:~# for i in "${!fields[@]}"; do
>   if [[ "${fields[$i]}" ]]; then
>     echo "field: ${i} | value: ${fields[$i]}"
>   fi
> done
field: python | value: inputs.python_version
field: runner | value: inputs.runner
```

## building args for the submit-thread action
```
root:~# sto=(--instance https://foo.bar --run-id 88 --results bar.xml)
root:~# echo "${sto[@]}"
--instance https://foo.bar --run-id 88 --results bar.xml
root:~# sto+=(--thread-resources foo_bar.xml)
root:~# echo "${sto[@]}"
--instance https://foo.bar --run-id 88 --results bar.xml --thread-resources foo_bar.xml
```

derekk-nm committed:
allow TestMo resources

add support for creating TestMo "resource" files and allow passing one to the testmo-run-submit-thread action

derekk-nm committed:
allow TestMo resources

add support for creating TestMo "resource" files and allow passing one to the testmo-run-submit-thread action

dbarbuzzi commented:
> looping through an associative array in the action that creates the resource file

If you have the time, it might be worth creating a small Python script to do this and see how it compares to Bash – while it's certainly possible in Bash as your changes indicate, it is likely much more readable/maintainable in Python (it can certainly be accomplished with plain Python3/no external libraries, so the action should still be reasonably portable).

dhuangnm commented on actions/testmo-run-submit-thread/action.yml:
```diff
@@ -39,15 +43,19 @@ runs:
-            --results ${RESULTS} \
-            -- ./.github/scripts/step-status ${{ inputs.step_status }}
+          submit_thread_options=( \
+            --instance ${inputs.testmo_url} \
```
Why does the testmo_url needs to be input now and if it has to be it seems not added as input yet?

derekk-nm replied:
line 4 defines this "testmo_url" input.
It is required because this command actually submits the thread to the TestMo server.

dbarbuzzi replied:
I think the question is more so why are we changing this instead of using the pre-existing `--instance ${TESTMO_URL}` / `--run-id ${TESTMO_RUN_ID}` / `--results ${RESULTS}` since the action already takes these inputs and transforms them into environment variables.

derekk-nm replied:
six of one, half dozen of the other?  It didn't seem appropriate to transform the new input into an env var, and I wanted to be consistent with usage w/in the bash code.  no big deal, I can change them back

dbarbuzzi replied:
I always like as few changes as makes sense. If you think it makes sense to keep the direct references to the `inputs` then by all means keep that, but I would then expect that the `env` block in the step is cleaned up if we're not going to use the env vars it declares (after making sure they aren't used anywhere else, of course).

derekk-nm replied:
understood.  i've reverted back to using the env vars.

dhuangnm commented on actions/testmo-create-resources/action.yml:
```diff
@@ -0,0 +1,38 @@
+  steps:
+    - run: |
+        echo "composing TESTMO resources file ${{ inputs.resources_file }} ..."
+        declare -A fields=( \
```
I think it's better to just pass in the test_config and construct the fields from the test_config so that we don't need to have separate input arguments for each field, plus we don't need python for the lm eval reporting etc.

derekk-nm replied:
I'm not sure what you mean by the "test_config", but I get what you're saying about too many input args.  I'm working on an alternative.
re: the python_version is being provided because that's one of the "resources" to be added to the TestMo thread.  no python is being run based on this field.

dhuangnm replied:
The "test_config" from nm-build-test.yml which contains info for python version, runner, config file.

We don't need to add python_version in the lm-eval thread, it's only needed for the test thread reporting.

derekk-nm replied:
so, `nm-build-test.yml` generates multiple **LM-EVAL** jobs based on the matrix of `lm-eval-configurations` passed to it.  each LM-EVAL job gets one of those configurations.  each lm-eval-configuration is a string that looks like this (for example):
```
'{"label":"gcp-k8s-l4-solo","config":"./.github/lm-eval-configs/solo-duo-small-smoke-models.txt","timeout":"300"}'
```
the job then breaks out the specifics from this lm-eval-configuration string into args that are passed to the `nm-lm-eval.yml` workflow.  (see https://github.com/neuralmagic/nm-vllm-ent/blob/207471a9f8fb072d601f308d0c25e8e915fa4ea0/.github/workflows/nm-build-test.yml#L147-L152)

So, by the time we get to the `nm-lm-eval.yml` workflow (where I am adding the steps create a resources file and to report the thread results), there is no single "test_config".  I'm working on replacement code to re-combine the relevant "resources" into a single arg to pass to `testmo-create-resources/action.yml`.

wrt the version of python.  my notes from our conversation in standup yesterday had this as one of the "resource" attributes to include w/ the LM-Eval.  If you don't think we want it after all, I'll remove it.

dbarbuzzi replied:
> wrt the version of python. my notes from our conversation in standup yesterday had this as one of the "resource" attributes to include w/ the LM-Eval. If you don't think we want it after all, I'll remove it.

I don’t think it’s necessary for `LM-EVAL` like it is for `TEST` since `LM-EVAL` all use a single Python (IIRC) while `TEST` has multiple runs for different Pythons. That said, having it on `LM-EVAL` doesn't hurt anything so I would vote that since it’s already there, leave it in.

dhuangnm replied:
Yes that's a good idea to recombine runner, config file etc into a single arg and pass to the create resource action. This way we only need to put in necessary fields into the arg to report for the thread resource. I still feel it's better not to include python info for lm-eval as this is not needed at all, and the simpler the better.

derekk-nm replied:
with my revisions, it's easy enough to add it back if we want it.

dhuangnm commented on actions/testmo-create-resources/action.yml:
```diff
@@ -0,0 +1,38 @@
+        for i in "${!fields[@]}"; do 
+          if [[ "${fields[$i]}" ]]; then
+            npx testmo automation:resources:add_field \
+              --resources "${{ inputs.resources_file }}" \
```
This should be for the --thread-resources not for the whole testmo run resources right?

derekk-nm replied:
this `npx testmo automation:resources:add_field` command is adding a field to a JSON file that will be passed to the `npx testmo automation:run:submit-thread` command via the `testmo-run-submit-thread/action.yml` file in the actions repo.  
The line here is pointing to the JSON file where the resources are to be stored.

dhuangnm replied:
I meant the resources should be for the --thread-resources as we're submitting results from Test and lm-eval jobs which has its own resource contents.

derekk-nm replied:
Well, the resources created with this action could be used for any `npx testmo automation:run:*` command.  for my nm-vllm-ent repo changes, I'll specifically be using the file w/ the TestMo run's thread, but that's not really relevant at this line in the code.

some links to the docs might help?

https://docs.testmo.com/docs/automation/reference#submitting-threads-for-runs
https://docs.testmo.com/docs/automation/reference#adding-fields-links-and-artifacts

dhuangnm replied:
The fields we need to add are only for threads not for the entire testmo run. As mentioned in Domenic's comment as well, we just need to add these fields info for the --thread-resources argument.

dbarbuzzi replied:
This command has no relation to threads or runs, though, it is simply building a resources file (which itself is independent of threads or runs). This command to build the file only has a `--resources` flag because those concepts are unrelated to building the resources, they are only related to actually submitting them to a thread/run.

derekk-nm replied:
Yup, that's being done in the other file in this PR.  The command in *this* file is to add fields to a resources file whose path is provided by the `--resources` arg on this line.

dhuangnm replied:
But those fields should only apply to thread right?

derekk-nm replied:
Dan and I discussed this after standup.  I believe that it's now clear that this `--resources` arg is simply the name of the resources file, and how the same file name is passed to the `--thread_resources` arg for testmo's `automation:run:submit-thread` command.  Marking this convo resolved.

dhuangnm commented on actions/testmo-create-resources/action.yml:
```diff
@@ -0,0 +1,38 @@
+name: create testmo resources
```
Where do we call this action from? Can you provide an example of calling it?

derekk-nm replied:
I'll be adding the following step to the `.github/workflows/nm-lm-eval.yml` file in the nm-vllm-ent repo:

```
      - name: create testmo resources
        id: create_testmo_resources
        uses: ./github/actions/nm-testmo-create-resources/
        if: (success() || failure()) && (inputs.testmo_run_id != '')
        with:
          resources_file: testmo_lmeval_resources.json
          python: ${{ inputs.python }}
          runner: ${{ inputs.test_label }}
          lm_eval_configuration: ${{ inputs.lm_eval_configuration }}
```

dhuangnm replied:
ok got it, again let's not pass in each field as argument but use the test_config and do the parsing in the nm-testmo-create-resources action. Also maybe we should rename the action to nm-testmo-create-thread-resources as it's not for the testmo run but for each thread.

derekk-nm replied:
I'll work on a way to pass in something other than hard coded fields.  That makes this action even more generic. 
 Any `testmo automation:run:*` command can accept a resources.json file, so it makes sense to leave the name as is.

dbarbuzzi commented on actions/testmo-run-submit-thread/action.yml:
```diff
@@ -13,6 +13,10 @@ inputs:
     description: "directory of JUnit '*.xml' formatted result files"
     required: true
+  thread_resources:
+    description: "XML file with TestMo resources to attach to this thread"
```
nit: It's JSON, not XML, unless something has drastically changed in the newer Testmo versions (though it certainly still accepts JSON as other jobs in other projects pass JSON).

dbarbuzzi commented on actions/testmo-run-submit-thread/action.yml:
```diff
@@ -39,15 +43,19 @@ runs:
           echo "submitting results to TESTMO run ..."
           ## not checking testmo_url and token as this should be
           ## called between "create" and "complete"
-          npx testmo automation:run:submit-thread \
```
As Dan mentioned, it seems like the only real change that would have been needed in this part of the action is to include the `--thread-resources` flag and value – that should have already been built prior to this action running by the job/workflow calling your new action prior to this one. Rather than trying to build CLI flags guard against no resources file being passed, it would be simpler to simply create an empty dummy resources file if one doesn't exist (that's what I have done in other projects): `cat {} > resources-file.json` is a minimal “empty” file that Testmo accepts.

derekk-nm replied:
Yes, the resources file would have been built before this action.  But, cool, I didn't expect TestMo would accept an empty resources file.  Since this action could still be called by a workflow that does not create the resource file, I'll still need to conditionally create the empty file in this action.

dbarbuzzi replied:
Yup, I did so here in a separate version of these actions for test-automation:

https://github.com/neuralmagic/test-automation/blob/main/.github/actions/testmo-run-submit-thread/action.yml#L40-L43

As you can see, it's a much simpler bit of logic that eliminates the need for more complex logic later on.

derekk-nm replied:
this is helpful.

dbarbuzzi reviewed (commented):
Left a couple of comments, let me know what you think.

derekk-nm committed:
PR updates

- don't use bash in the create-resources action, instead use a python script to do the loop calling the testmo automation:resources:add-field command
- pass around resource information as a JSON associative array
- revert to using env vars instead of input variables in the submit-thread action
- don't compose the submit-thread args.  since the resources file can be an empty JSON, simply add that option to the command and pass in an empty file if one wasn't provided.

derekk-nm committed:
PR updates

- don't use bash in the create-resources action, instead use a python script to do the loop calling the testmo automation:resources:add-field command
- pass around resource information as a JSON associative array
- revert to using env vars instead of input variables in the submit-thread action
- don't compose the submit-thread args.  since the resources file can be an empty JSON, simply add that option to the command and pass in an empty file if one wasn't provided.

derekk-nm commented:
@dhuangnm and @dbarbuzzi , I've completed updates based on your PR feedback.  You can see the corresponding changes in the nm-vllm-ent repo in my [dmk/try_lm-eval_to_TestMo](https://github.com/neuralmagic/nm-vllm-ent/tree/dmk/try_lm-eval_to_TestMo) branch.
sadly, there seems to be a problem w/ the way that I've referred to the actions in this repo.  I'll have to address that when I return from PTO.  In the meantime, I'd appreciate it if you would take a look at the changes here.
For further context, you can see these actions being used in https://github.com/neuralmagic/nm-vllm-ent/compare/main...dmk/try_lm-eval_to_TestMo#diff-a5bdf4dabbb4859c7d32b9553da3c141baed531be90b5c117dfd578f34d9a829 and https://github.com/neuralmagic/nm-vllm-ent/compare/main...dmk/try_lm-eval_to_TestMo#diff-c7b45298d29a28cedc608402670c800b3d180cc8c5a3529c3e141393331bb659

derekk-nm committed:
Merge branch 'main' into dmk/testmo_resources

derekk-nm committed:
Merge remote-tracking branch 'origin/dmk/testmo_resources' into dmk/testmo_resources

# Conflicts:
#	actions/testmo-run-submit-thread/action.yml

derekk-nm committed:
remove extra )

derekk-nm committed:
additional fixes

create-resource action needs to add the actions path to PATH and quote the option values.
update the compose_testmo_resources_file.py to correct some typos and pass the "shell=True" option to the subprocess.run() command.
 [thanks dbarbuzzi]

derekk-nm committed:
quote automation:run:submit-thread command option values

derekk-nm committed:
mark the script executable

derekk-nm committed:
fix typo

derekk-nm committed:
correctly submit testmo resources with submit-thread

derekk-nm committed:
Merge branch 'main' into dmk/testmo_resources

derekk-nm commented:
There were so many conflicts w/ recent updates to this repo that it was cleaner to create a new PR.  Closing this in favor of https://github.com/neuralmagic/nm-actions/pull/32
